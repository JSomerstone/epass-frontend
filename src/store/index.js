import Vue from "vue";
import Vuex from "vuex";

import axios from "axios";
import config from "../config/index";

import tournaments from "./modules/tournaments";
import referees from "./modules/referees";
import countries from "./modules/countries";
import auth from "./modules/auth";

axios.defaults.baseURL = config.backendUrl;
Vue.use(Vuex);

const store = new Vuex.Store({
  modules: {
    tournaments,
    referees,
    countries,
    auth
  }
});

export default store;