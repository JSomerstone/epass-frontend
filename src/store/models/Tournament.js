import Comment from "./Comment";

export default class Tournament {
  constructor(rawData = {}, refereeList = []) {
    const {
      id = null,
      name = "",
      international = true,
      city = "",
      country = "",
      year = null,
      dates = [],
      td = {},
      referees = [],
      teams = [],
      comments = { items: [] },
      createdBy = null,
    } = rawData;

    this.id = id;
    this.international = international;
    this.name = name;
    this.city = city;
    this.country = country;
    this.dates = dates.map(d => new Date(d));
    this.year = year || new Date().getFullYear();
    this.td = typeof td == "string" ? this.getRef(td, refereeList) : td;
    this.referees = referees.map(
      r => {
        return {
          ...this.getRef(r.id, refereeList),
          ...r
        }
      }
    );
    this.comments = {
      items: comments.items.map(c => new Comment(c))
    }
    this.teams = teams;
    this.createdBy = createdBy;
  }
  getRef(id, refereeList) {
    return refereeList.find(
      r => r.id == id
    ) || {};
  }
  
  setGames(refereeId, games, tenSeconds) {
    const index = this.referees.findIndex(
      r => r.id === refereeId
    )
    if (index < 0) {
      return;
    }
    Object.assign(
      this.referees[index],
      { games, tenSeconds }
    );
  }
  toJson() {
    return {
      id: this.id,
      international: this.international,
      name: this.name,
      city: this.city,
      country: this.country,
      dates: this.dates.map(d => d.toUTCString()),
      year: this.year,
      td: this.td.id,
      referees: this.referees.map(
        r => {
          const { id, games = 0, tenSeconds = 0 } = r;
          return { id, games, tenSeconds };
        }
      ),
      teams: this.teams,
      createdBy: this.createdBy
    };
  }

  validate() {
    const validity = {
      international: typeof this.international == "boolean",
      name: this.name.length > 2,
      city: this.city.length >= 2,
      country: this.country != "",
      dates: this.dates.length == 2,
      year: 2019 <= this.year && this.year <= new Date().getFullYear()+1
    };
    const errors = Object.entries(validity)
      .filter(entry => !entry[1])
      .map(entry => entry[0]);
    return errors;
  }

  isValid() {
    return this.validate().length == 0;
  }
}
