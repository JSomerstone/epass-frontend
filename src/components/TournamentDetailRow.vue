<template>
  <div class="columns">
    <div class="column is-one-third">
      <b-button outline type="is-info" @click="openEditForm" icon-left="file-document-edit-outline">
        Edit
      </b-button>
      <div v-if="hasGames">
        <strong>Your statistics:</strong>
        <ul>
            <li>{{ statistics.games }} Game(s) as referee</li>
            <li>{{ statistics.games }} Game(s) as table official / 10 sec timer</li>
        </ul>
      </div>
    </div>
    <div class="column is-one-third">
      <div>
        <strong>TD:</strong> {{ getRefereeName(tournament.td) }}
      </div>
      <div>
        <strong>Referees:</strong>
        <ul>
          <li v-for="referee in tournament.referees" v-bind:key="referee.id">
            {{ getRefereeName(referee.id) }}
          </li>
        </ul>
      </div>
    </div>
    <div class="column is-one-third">
      <strong>Teams</strong>
      <ul>
        <li v-for="team in tournament.teams" v-bind:key="team">
          {{ team }}
        </li>
      </ul>
    </div>
  </div>
</template>
<script>
import Tournament from "../store/models/Tournament"
export default {
  props: {
    tournament: {
      type: Object,
      default: () => new Tournament()
    }
  },
  computed: {
    statistics: function() {
      const { games= 0, tenSeconds= 0} = this.getCurrentRef() || {};
      return { games, tenSeconds }
    },
    hasGames: function() {
      const ref = this.getCurrentRef();
      return ref 
        ? ref.games || ref.tenSeconds
        : false;
    }
  },
  methods: {
    getRefereeName: function(id) {
      const ref = this.$store.getters['referees/byId'](id);
      return `${ref.firstName} ${ref.lastName} [${ref.country}]`;
    },
    getCurrentRef: function() {
      const current = this.$store.getters['referees/current'];
      if (! current) {
        return;
      }
      return this.tournament.referees.find( r => r.id == current.id );
    },
    openEditForm: function() {
      this.$store.dispatch("tournaments/openForEdit", { tournament: this.tournament });
    }
  },
}
</script>