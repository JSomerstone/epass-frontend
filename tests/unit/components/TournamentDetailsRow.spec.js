import { mount, RouterLinkStub } from "@vue/test-utils";
import Vuex from "vuex";
import localVue from "../localVue";
import store from "../mockStore";
import testReferees from "../../data/referee.json";

import TournamentDetailRow from "@/components/TournamentDetailRow";
import Tournament from "@/store/models/Tournament";


describe("TournamentDetailRow", () => {
  let tournament, referees = [];
  
  beforeEach(() => {
    referees = testReferees.map(
      r => {
        return {
          id: r.id,
          games: 6,
          tenSeconds: 7
        };
      }
    )
    tournament = new Tournament(
      {
        id: "bogus-tournament",
        td: testReferees[0].id,
        referees,
        teams: ["Team A", "Team B"]
      },
      testReferees
    ).toJson();
  });

  it("Renders correctly", () => {
    const wrapper = mount(TournamentDetailRow, {
      store: new Vuex.Store(store),
      localVue,
      propsData: { tournament },
      stubs: {
        RouterLink: RouterLinkStub
      }
    });

    expect(wrapper).toMatchSnapshot();
  })
});