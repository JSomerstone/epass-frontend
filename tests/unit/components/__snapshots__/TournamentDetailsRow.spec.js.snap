// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TournamentDetailRow Doesn't render users statistics when user is not a referee 1`] = `
<div class="columns">
  <div class="column is-one-third"><a><button type="button" outline="" class="button is-info"><span class="icon is-small"><i class="mdi mdi-folder-open-outline"></i></span> <span>
        Open
      </span>
        <!----></button></a>
    <!---->
  </div>
  <div class="column is-one-third">
    <div><strong>TD:</strong> John Doe [USA]
    </div>
    <div><strong>Referees:</strong>
      <ul>
        <li>
          Jane Doe [CAN]
        </li>
        <li>
          Noob Nederson [SWE]
        </li>
      </ul>
    </div>
  </div>
  <div class="column is-one-third"><strong>Teams</strong>
    <ul>
      <li>
        Team A
      </li>
      <li>
        Team B
      </li>
    </ul>
  </div>
</div>
`;

exports[`TournamentDetailRow Renders correctly 1`] = `
<div class="columns">
  <div class="column is-one-third"><a><button type="button" outline="" class="button is-info"><span class="icon is-small"><i class="mdi mdi-folder-open-outline"></i></span> <span>
        Open
      </span>
        <!----></button></a>
    <div><strong>Your statistics:</strong>
      <ul>
        <li>
          6
          Games as referee
        </li>
        <li>
          7
          Games
          as table official / 10 sec timer
        </li>
      </ul>
    </div>
  </div>
  <div class="column is-one-third">
    <div><strong>TD:</strong> John Doe [USA]
    </div>
    <div><strong>Referees:</strong>
      <ul>
        <li>
          John Doe [USA]
        </li>
        <li>
          Jane Doe [CAN]
        </li>
        <li>
          Noob Nederson [SWE]
        </li>
      </ul>
    </div>
  </div>
  <div class="column is-one-third"><strong>Teams</strong>
    <ul>
      <li>
        Team A
      </li>
      <li>
        Team B
      </li>
    </ul>
  </div>
</div>
`;
